### ISSUE TITLE
Async actions don't work using TypeScript and `es5` target

### ISSUE THREAD

**Issue by praxxis** [mobxjs/mobx-state-tree#279]
Per https://github.com/Microsoft/TypeScript/issues/15088, TS using `"target": "es5"` doesn't brand a generator function with a proper `name` or `displayName`, which causes [`isGeneratorFunction`](https://github.com/mobxjs/mobx-state-tree/blob/master/src/utils.ts#L91-L97) to return a false negative. 

Furthermore, the MST [tests tsconfig](https://github.com/mobxjs/mobx-state-tree/blob/master/test/tsconfig.json#L5) is set to the `es2015` target, which is why this is currently passing in tests (the base tsconfig is set to `es5`).

TBH I'm not sure what a good fix for this is. Checking an action for a `next` function or some such might work, but feels wrong. There's nothing else in the TS [`__generator` polyfill ](https://github.com/Microsoft/tslib/blob/master/tslib.es6.js#L69-L95) that could be used to discover a generator function.

**Comment by KatSick** [comment]
+1 . blocker for me

**Comment by mweststrate** [comment]
Sorry guys, nasty oversight. Is this something that needs to be fixed urgently? #282 will address it properly, but otherwise we could introduce an `async(function* () {})` method that marks the method properly as work around. (it screws typings a bit though). Basically undoing this [commit](https://github.com/mobxjs/mobx-state-tree/commit/5469bd2f44c0c1d1c8ec7529ecf03d8036ac5003). But instead of check for `isGenerator`, explicitly marking it as generator.

**Comment by KatSick** [comment]
hi @mweststrate. not urgent at all. Helper `async(function * () {})` is great workaround except for Typescript users. With `strictImplicitAny: true` all `this` inside actual generator `function * () {}` will be any..

the only wild idea how to fix it (i think we should NOT do this), is to extend checking in [isGeneratorFunction](https://github.com/mobxjs/mobx-state-tree/blob/master/src/utils.ts#L91-L97) to support weird Typescript bugs... (here is link to playground [TS playground](https://www.typescriptlang.org/play/index.html#src=function%20isGeneratorFunction(value%3A%20any)%20%7B%0A%20%20%20%20const%20constructor%20%3D%20value.constructor%3B%0A%0A%20%20%20%20if%20(!constructor)%20%7B%0A%20%20%20%20%20%20%20%20return%20false%3B%0A%20%20%20%20%7D%0A%20%20%20%20if%20(%22GeneratorFunction%22%20%3D%3D%3D%20constructor.name%20%7C%7C%20%22GeneratorFunction%22%20%3D%3D%3D%20constructor.displayName%20%7C%7C%20value.toString().indexOf('return%20__generator')%20!%3D%3D%20-1)%20%7B%0A%20%20%20%20%20%20%20%20return%20true%3B%0A%20%20%20%20%7D%0A%0A%20%20%20%20return%20false%3B%0A%7D%0A%0Afunction*%20typescriptsometimessucks()%20%7B%0A%20%20%20%20return%20null%3B%0A%7D%0A%0Aconsole.log(isGeneratorFunction(typescriptsometimessucks))%3B) with this fix: )
```
export function isGeneratorFunction(value: any) {
    const constructor = value.constructor
    if (!constructor) return false
    if ("GeneratorFunction" === constructor.name || "GeneratorFunction" === constructor.displayName) || value.toString().indexOf('return __generator') !== -1
        return true
    return false
}

```

**Comment by mattiamanzati** [comment]
This has been fixed with the release of 0.10 which introduces a new async API that provides correct typings and works with transpiled generators.
